test/modules/TestResultCoordinator/Reports/DeploymentTestReportGenerator.cs

```diff
@@ -0,0 +1,156 @@
+            this.ExpectedSource = Preconditions.CheckNonWhiteSpace(expectedSource, nameof(expectedSource));
+            this.ActualSource = Preconditions.CheckNonWhiteSpace(actualSource, nameof(actualSource));
+            this.ActualTestResults = Preconditions.CheckNotNull(actualTestResults, nameof(actualTestResults));
+
+            this.TestResultComparer = new DeploymentTestResultComparer();
+            this.ResultType = TestOperationResultType.Deployment.ToString();
+        }
+
+        internal string ActualSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> ActualTestResults { get; }
+
+        internal string ExpectedSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> ExpectedTestResults { get; }
+
+        internal string ResultType { get; }
+
+        internal ITestResultComparer<TestOperationResult> TestResultComparer { get; }
+
+        /// <summary>
+        /// Compare 2 data stores and counting expected, actual, and matched results; and return a deployment test report.
+        /// Actual deployment results can be less than expected deployment results.
+        /// An actaul deployment test result is possible to use for verification of more than 1 expected deplpyment test result.
+        /// It will log fail if actual store has more results than expect store.
+        /// </summary>
+        /// <returns>Test Result Report.</returns>
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            Logger.LogInformation($"Start to generate report by {nameof(DeploymentTestReportGenerator)} for Sources [{this.ExpectedSource}] and [{this.ActualSource}]");
+
+            var lastActualDeploymentTestResult = default(TestOperationResult);
+            ulong totalExpectedDeployments = 0;
+            ulong totalActualDeployments = 0;
+            ulong totalMatchedDeployments = 0;
+            var unmatchedResults = new List<TestOperationResult>();
+
+            bool hasExpectedResult = await this.ExpectedTestResults.MoveNextAsync();
+            if (hasExpectedResult)
+            {
+                totalExpectedDeployments++;
+            }
+
+            bool hasActualResult = await this.ActualTestResults.MoveNextAsync();
+            if (hasActualResult)
+            {
+                lastActualDeploymentTestResult = this.ActualTestResults.Current;
+                totalActualDeployments++;
+            }
+
+            while (hasExpectedResult && hasActualResult)
+            {
+                this.ValidateDataSource(this.ExpectedTestResults.Current, this.ExpectedSource);
+                this.ValidateDataSource(this.ActualTestResults.Current, this.ActualSource);
+
+                if (this.TestResultComparer.Matches(this.ExpectedTestResults.Current, this.ActualTestResults.Current))
+                {
+                    totalMatchedDeployments++;
+
+                    hasExpectedResult = await this.ExpectedTestResults.MoveNextAsync();
```

Comment:
if they match the actual results should also be moved to next value

Reply:
Explained in person